cmake/external_protobuf.cmake

```diff
@@ -0,0 +1,121 @@
+        INSTALL_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf"
+        EXCLUDE_FROM_ALL TRUE
+        )
+else()
+    if (${CMAKE_VERSION} VERSION_LESS 3.6)
+        ExternalProject_Add(
+            ext_protobuf
+            PREFIX ext_protobuf
+            GIT_REPOSITORY ${PROTOBUF_GIT_REPO_URL}
+            GIT_TAG ${PROTOBUF_GIT_BRANCH}
+            INSTALL_COMMAND ""
+            UPDATE_COMMAND ""
+            PATCH_COMMAND ""
+            CONFIGURE_COMMAND ./autogen.sh && ./configure --disable-shared CXXFLAGS=-fPIC
+            BUILD_COMMAND ${MAKE}
+            TMP_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/tmp"
+            STAMP_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/stamp"
+            DOWNLOAD_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/download"
+            SOURCE_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/src"
+            BINARY_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/src"
+            INSTALL_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf"
+            BUILD_BYPRODUCTS "${EXTERNAL_PROJECTS_ROOT}/protobuf/src/src/.libs/libprotobuf.a"
+            EXCLUDE_FROM_ALL TRUE
+            )
+    else()
+        # To speed things up prefer 'shallow copy' for CMake 3.6 and later
+        ExternalProject_Add(
+            ext_protobuf
+            PREFIX ext_protobuf
+            GIT_REPOSITORY ${PROTOBUF_GIT_REPO_URL}
+            GIT_TAG ${PROTOBUF_GIT_BRANCH}
+            GIT_SHALLOW TRUE
+            INSTALL_COMMAND ""
+            UPDATE_COMMAND ""
+            PATCH_COMMAND ""
+            CONFIGURE_COMMAND ./autogen.sh && ./configure --disable-shared CXXFLAGS=-fPIC
+            BUILD_COMMAND ${MAKE}
+            TMP_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/tmp"
+            STAMP_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/stamp"
+            DOWNLOAD_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/download"
+            SOURCE_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/src"
+            BINARY_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf/src"
+            INSTALL_DIR "${EXTERNAL_PROJECTS_ROOT}/protobuf"
+            BUILD_BYPRODUCTS "${EXTERNAL_PROJECTS_ROOT}/protobuf/src/src/.libs/libprotobuf.a"
+            EXCLUDE_FROM_ALL TRUE
+            )
+    endif()
+endif()
+
+# -----------------------------------------------------------------------------
+
+ExternalProject_Get_Property(ext_protobuf SOURCE_DIR BINARY_DIR)
+
+# -----------------------------------------------------------------------------
+# Use the interface of FindProtobuf.cmake
+# -----------------------------------------------------------------------------
+
+set(Protobuf_SRC_ROOT_FOLDER ${SOURCE_DIR})
+
+set(Protobuf_PROTOC_EXECUTABLE ${Protobuf_SRC_ROOT_FOLDER}/src/protoc)
+set(Protobuf_INCLUDE_DIR ${Protobuf_SRC_ROOT_FOLDER}/src)
+set(Protobuf_LIBRARY ${Protobuf_SRC_ROOT_FOLDER}/src/.libs/libprotobuf.a)
+set(Protobuf_LIBRARIES ${Protobuf_LIBRARY})
+
+if (NOT TARGET protobuf::libprotobuf)
+    add_library(protobuf::libprotobuf UNKNOWN IMPORTED)
+    set_target_properties(protobuf::libprotobuf PROPERTIES
```

Comment:
use target_include_directories like this https://github.com/NervanaSystems/ngraph/blob/master/cmake/external_gtest.cmake#L77

Reply:
It will result in error message (see below). The "namespace::target" semantics needs 'set_target_properties' instead. I use the same API FindProtobuf module is exposing.
```
CMake Error at cmake/external_protobuf.cmake:110 (target_include_directories):
  target_include_directories called with non-compilable target type
```